### PR TITLE
Add AOT building

### PR DESCRIPTION
**Pull request by jld3103**
Using AOT engine builds from https://github.com/flutter-rs/engine-builds.
Also see https://github.com/go-flutter-desktop/go-flutter/issues/331
Test that AOT works:
- [x] Linux
- [ ] Windows
- [ ] Darwin

1. Install hover AOT using `git clone https://github.com/jld3103/hover -b feature/aot`. Then go into `hover` and run `go install`.
2. Make sure `hover` is in your PATH. If not find a tutorial for your platform.
3. Run `hover build PLATFORM --release` where PLATFORM is your host OS (`linux`, `windows`, `darwin`)
4. Execute the resulting executable in `./go/build/outputs/PLATFORM/` (Preferrably in a terminal to catch any errors)
5. Report if it worked for you on your platform. Please also attach a zip of `./go/build/outputs/PLATFORM/` if something failed (building or running). Linux is already proven to work. Mac and Windows should work, but have to be tested.

### CODE REVIEW

File: cmd/build.go [hunk 1]
```diff
@@ -424,21 +431,99 @@ func buildNormal(targetOS string, vmArguments []string) {
...
+			if err != nil {
+				log.Errorf("Generating kernel snapshot failed: %v", err)
+				os.Exit(1)
+			}
+			cmdGenerateAotSnapshot := exec.Command(
+				genSnapshot,
+				"--causal_async_stacks",
+				"--deterministic",
+				"--snapshot_kind=app-aot-elf",
+				"--strip",
```
Comment (pchampio) [inline cmd/build.go]: Do we want to `--strip` in Profile mode?

File: cmd/build.go [hunk 2]
```diff
@@ -424,21 +431,99 @@ func buildNormal(targetOS string, vmArguments []string) {
...
+			}
+			err = os.Remove(filepath.Join(build.OutputDirectoryPath(targetOS), "flutter_assets", "kernel_blob.bin"))
+			if err != nil {
+				log.Errorf("Failed to remove unused kernel_blob.bin: %v", err)
+				os.Exit(1)
+			}
+			dart := filepath.Join(engineCachePath, "dart"+build.ExecutableExtension(targetOS))
+			genSnapshot := filepath.Join(engineCachePath, "gen_snapshot"+build.ExecutableExtension(targetOS))
+			kernelSnapshot := filepath.Join(build.OutputDirectoryPath(targetOS), "kernel_snapshot.dill")
+			elfSnapshot := filepath.Join(build.OutputDirectoryPath(targetOS), "libapp.so")
```
Comment (pchampio) [inline cmd/build.go]: I'm guessing .so because we are on linux. Maybe add a `TODO` saying we need to support other platform?

File: cmd/build.go [hunk 3]
```diff
@@ -609,11 +698,14 @@ func buildCommand(targetOS string, vmArguments []string, outputBinaryPath string
 		"-X github.com/go-flutter-desktop/go-flutter.ProjectVersion=%s "+
 			" -X github.com/go-flutter-desktop/go-flutter.PlatformVersion=%s "+
 			" -X github.com/go-flutter-desktop/go-flutter.ProjectName=%s "+
-			" -X github.com/go-flutter-desktop/go-flutter.ProjectOrganizationName=%s",
+			" -X github.com/go-flutter-desktop/go-flutter.ProjectOrganizationName=%s "+
+			" -X github.com/go-flutter-desktop/go-flutter.IsAot=%t ",
```
Comment (pchampio) [inline cmd/build.go]: `IsAot` can be removed

File: internal/enginecache/cache.go [hunk 4]
```diff
@@ -360,24 +376,80 @@ func ValidateOrUpdateEngineAtPath(targetOS string, cachePath string) (engineCach
...
-		err := moveFile(
-			filepath.Join(engineExtractPath, "libflutter_engine.so"),
-			filepath.Join(engineCachePath, "/libflutter_engine.so"),
-		)
+	// Strip linux JIT engine after download and not at every build
+	if targetOS == "linux" && !mode.IsAot {
+		err = exec.Command("strip", "-s", filepath.Join(engineCachePath, build.EngineFiles(targetOS, mode)[0])).Run()
 		if err != nil {
-			log.Errorf("Failed to move downloaded libflutter_engine.so: %v", err)
+			log.Errorf("Failed to strip %s: %v", filepath.Join(engineCachePath, build.EngineFiles(targetOS, mode)[0]), err)
```
Comment (pchampio) [inline internal/enginecache/cache.go]: `filepath.Join(engineCachePath, build.EngineFiles(targetOS, mode)[0])` can be placed in a local var

File: internal/enginecache/cache.go [hunk 5]
```diff
@@ -8,24 +8,28 @@ import (
...
+	"github.com/go-flutter-desktop/hover/internal/build"
 	"github.com/go-flutter-desktop/hover/internal/log"
-	"github.com/pkg/errors"
 )
 
 func createSymLink(oldname, newname string) error {
 	err := os.Remove(newname)
 	if err != nil && !os.IsNotExist(err) {
-		return errors.Wrap(err, "failed to remove existing symlink")
+		return Wrap(err, "failed to remove existing symlink")
```
Comment (pchampio) [inline internal/enginecache/cache.go]: Nit: leave the package name (`errors`) in front of the methods.

File: cmd/build.go (hunk 2 above)
Comment (jld3103) [reply cmd/build.go]: It's always an .so. They are used platform indpendent and are only the way of how the snapshots are bundled

File: internal/enginecache/cache.go (hunk 5 above)
Comment (jld3103) [reply internal/enginecache/cache.go]: That was some refactoring done by the IDE. It wasn't intentionally

File: cmd/build.go (hunk 1 above)
Comment (jld3103) [reply cmd/build.go]: Also some other flags need to be changes for profile mode. Can we ask the flutter team about that? In the mean time I would simply remove `--strip` for profile mode.

File: internal/enginecache/cache.go [hunk 6]
```diff
@@ -360,24 +376,81 @@ func ValidateOrUpdateEngineAtPath(targetOS string, cachePath string) (engineCach
...
+	// Strip linux JIT engine after download and not at every build
+	if targetOS == "linux" && !mode.IsAot {
+		unstrippedEngineFile := filepath.Join(engineCachePath, build.EngineFiles(targetOS, mode)[0])
+		err = exec.Command("strip", "-s", unstrippedEngineFile).Run()
 		if err != nil {
-			log.Errorf("Failed to move downloaded libflutter_engine.so: %v", err)
+			log.Errorf("Failed to strip %s: %v", unstrippedEngineFile, err)
 			os.Exit(1)
 		}
+	}
```
Comment (pchampio) [inline internal/enginecache/cache.go]: The linux AOT `libflutter_engine.so` file can also be strip-ed (40M -> 12M).

Comment (jld3103) [reply internal/enginecache/cache.go]: That exactly is what this piece of code does

Comment (jld3103) [reply internal/enginecache/cache.go]: Interesting. Going to implement that

**Comment by jld3103** [comment]
Currently doesn't work on darwin, because of the required `FlutterEmbedder` framework (https://github.com/go-flutter-desktop/go-flutter/blob/bfdc2df08a17d6ce3cc28df02700fc0ea535be9a/embedder/build.go#L5)

**Comment by Tokenyet** [comment]
In windows:
```
hover: Downloading AOT release engine for platform windows-x64 at version a67792536c...
Download completed in 203.10s
hover: Failed to move downloaded flutter_engine.dll.exp: Couldn't open src file: open C:\Users\MyAccount\AppData\Local\Temp\hover-engine-download237242011\engine\flutter_engine.dll.exp: The system cannot find the file specified.
```
I thought It's permission problem, so I tried with *open with administrator*, but still got the same message.

In the folder, there is a file `engine.zip` and a folder `engine`, and here is the structure list from the folder (`dir`):
```
2020/02/11  下午 08:58    <DIR>          .
2020/02/11  下午 08:58    <DIR>          ..
2020/02/11  下午 08:58        22,655,488 dart.exe
2020/02/11  下午 08:58            81,099 flutter_embedder.h
2020/02/11  下午 08:58             6,365 flutter_engine.exp
2020/02/11  下午 08:58            11,240 flutter_engine.lib
2020/02/11  下午 08:58       104,001,536 flutter_engine.pdb
2020/02/11  下午 08:58    <DIR>          flutter_patched_sdk
2020/02/11  下午 08:58    <DIR>          gen
2020/02/11  下午 08:58         3,834,368 gen_snapshot.exe
2020/02/11  下午 08:58           901,808 icudtl.dat
               7 個檔案     131,491,904 位元組
               4 個目錄   3,701,338,112 位元組可用
```

Hope these help!

**Comment by jld3103** [comment]
The engine builds changed file names. I'll fix them in an hour

**Comment by jld3103** [comment]
Copying of the files should work now, but I'm still not sure if windows works at all, but I have hopes that it works now.

**Comment by jld3103** [comment]
Hm trying it on my windows 10 VM I run into `error adding symbols: File in wrong format`. This seems like an issue with the compile, because the engine is an MSVC library, but TDM-GCC expects another type. We can either try MSVC or mingw64. I guess mingw64 would be easier for most people. @Tokenyet do you have any of these installed? If so could you help testing?

**Comment by jld3103** [comment]
I'll try compiling and running using wine which could maybe work and if so it would be a greate achievement, because we could produce cross-compiled AOT builds.

**Comment by Tokenyet** [comment]
I got the same error:

```
ror adding symbols: File in wrong format
collect2.exe: error: ld returned 1 exit status
github.com/go-flutter-desktop/go-flutter/internal/opengl
hover: Go build failed: exit status 2
```
In additional, I used TDM-GCC-64, and I'm glad to help this in my free time 😄

**Comment by Tokenyet** [comment]
@jld3103 , I tried Mingw64, here is the log:
```
hover: Bundling flutter app
hover: Generating kernel snapshot
hover: Generating ELF snapshot
Warning: Generating ELF library without DWARF debugging information.
hover: Failed to retrieve the organization name: open android/app/src/main/AndroidManifest.xml: The system cannot find the path specified.
hover: Compiling 'go-flutter' and plugins
hover: Successfully compiled
```

But `flutter_engine.dll` is missing from the final folder, so It popup alert dialog. To clearify, I change back to original hover,  `flutter_engine.dll` is generated.

Mingw64 with aot
```
2020/02/12  下午 05:24    <DIR>          .
2020/02/12  下午 05:24    <DIR>          ..
2020/02/12  下午 05:24    <DIR>          assets
2020/02/12  下午 05:24         9,425,891 example_flutter.exe
2020/02/12  下午 05:23    <DIR>          flutter_assets
2020/02/12  下午 05:24             6,365 flutter_engine.exp
2020/02/12  下午 05:24            11,240 flutter_engine.lib
2020/02/12  下午 05:24       104,001,536 flutter_engine.pdb
2020/02/12  下午 05:24           901,808 icudtl.dat
2020/02/12  下午 05:24         5,096,288 libapp.so
```

Original Structure:
```
2020/02/12  下午 05:19    <DIR>          .
2020/02/12  下午 05:19    <DIR>          ..
2020/02/12  下午 05:19    <DIR>          assets
2020/02/12  下午 05:19         3,457,024 example_flutter.exe
2020/02/12  下午 05:19    <DIR>          flutter_assets
2020/02/12  下午 05:19        25,208,832 flutter_engine.dll
2020/02/12  下午 05:19           901,808 icudtl.dat
               3 個檔案      29,567,664 位元組
               4 個目錄  78,756,683,776 位元組可用
```

**Comment by jld3103** [comment]
That's right. There should be no dll file in the output folder. Did you try running the build?

**Comment by Tokenyet** [comment]
image

What I mean is the executable still ask for `flutter_engine.dll` 😢

**Comment by jld3103** [comment]
Ah hm

**Comment by jld3103** [comment]
@Tokenyet now it shouldn't ask for a `.dll` anymore. You need to clear the engine cache before trying again (the cache path is somewhere in `%APPDATA%\Local\hover`. I hope that's the right path on windows).

**Comment by jld3103** [comment]
@Tokenyet thanks for trying this PR and all the trial and error around it

**Comment by Tokenyet** [comment]
Now, dll is generated, but It got another issue 😕 

image

```
[FATAL:D:\a\engine-builds\engine-builds\engine\src\flutter\runtime\dart_vm.cc(421)] Error while initializing the Dart VM: Precompiled runtime requires a precompiled snapshot
```

**Comment by jld3103** [comment]
I've run into the issue before and I don't know how to fix it or even where the issue is. I guess it's in the engine builds.

**Comment by pchampio** [comment]
For later: https://github.com/flutter/flutter/blob/beta/packages/flutter_tools/lib/src/base/build.dart#L103-L303

**Comment by pchampio** [comment]
[ELF.zip](https://github.com/go-flutter-desktop/hover/files/4579860/ELF.zip)
Golang version of:
https://github.com/jwinarske/flutter_wayland/pull/1/files
https://github.com/flutter-rs/flutter-rs/commit/307bf7824bcdbc694de08d0b53db344e19cea1a5

As https://github.com/flutter/engine/pull/13324/files#r338745201

**Comment by jld3103** [comment]
I think this PR should be dropped, because it has many conflicts and operates in a completely different way than hover currently does with the new docker implementation.

**Comment by pchampio** [comment]
Ok for dropping it. But the issue can be leaved open. at least until we have another impl

**Comment by GeertJohan** [comment]
I guess we could keep it open as WIP / Draft, for future reference?

**Pull request closed without merging**